[1단계 - 행운의 로또 미션] 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
# 🍊 `Step1` 니야와 주렁의 페어 프로그래밍 🍊
### 안녕하세요, 동키콩! 니야입니다. 행운의 로또 미션 1단계 PR 제출하겠습니다!

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [ ] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [ ] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입:

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

TDD를 적용하면서 테스트 코드 → 프로덕션 코드 순서로 구현해야 했지만, 테스트 코드를 작성하기 위해 먼저 머릿속으로 프로덕션 코드를 구상하는 경우가 많아 TDD를 온전히 실천했다고 보기 어려웠던 것 같습니다.  
그러나 최대한 프로덕션 코드의 구현을 배제하고 테스트 코드를 우선적으로 작성하려고 노력하면서, 테스트 코드 덕분에 실시간으로 정확한 검증을 할 수 있다는 점을 배울 수 있었습니다. 

이번 TDD 과정에서는 테스트 코드 개수가 이전보다 많아졌는데, 이는 도메인 로직을 더 작은 단위로 분리하고 테스트했기 때문이라고 생각합니다. 덕분에 함수 하나가 하나의 역할을 정확히 수행하는지 철저히 검증할 수 있는 계기가 되었습니다.

TDD를 처음 하다보니 테스트 코드 작성하는데 시간이 더 오래걸린 느낌이 들었습니다. 사실 TDD는 모든걸 완벽하게 구상하여 테스트를 진행할 수 없기 때문에 시간을 줄이기 위해 하나하나 테스트 케이스를 작성해보며 그 로직을 점차 확장해나가는 식으로 구현하며 연습했습니다. 덕분에 갈수록 테스트 코드 구현할 때 시간이 단축됨을 느꼈습니다.

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

✅ **기본 질문**

- 순수함수가 아니라 더 분리할 필요가 있거나, 더 효율적으로 구현할 수 있는 방식이 있는 부분에 대해서 피드백 부탁드립니다.
- 테스트에서도 놓친 테스트 케이스가 있거나, 테스트 명세 부분에서 수정해야할 필요가 있는 부분이 있다면 말씀 부탁드립니다.

✅ **TDD에 관하여**

막상 진행해보니 시간이 생각보다 오래 걸리는 문제 때문에 궁금한 점이 생겼는데요, 마감 기한을 정확히 지켜야하는 실제 실무 상황에서도 TDD를 많이 적용하는 편인지 궁금합니다.
TDD를 하면 도메인 로직을 미리 머릿속으로 구상하고 테스트 코드를 작성하는 데도, 계속 코드 구조가 바뀌어서 테스트 코드를 다시 리팩토링해야 하는 경우가 발생합니다.
사실 그 과정이 번거롭긴 하지만, 그래도 아주 작은 단위로 쪼개어 로직을 구현하면 나름 괜찮을 것 같다는 생각이 들었습니다.
하지만, 꼭 TDD를 레드-그린 리팩토링 방식으로 해야하는지 의문이 들었습니다.
도메인을 구현하면서 그 작은 기능 단위에 대해 동시에, 또는 도메인 로직 구현 후에 테스트 코드를 작성하는 방식이 더 효율적인 느낌이 들었습니다.
따라서, 추가로 TDD의 효율 및 유용성에 대해서 어떻게 생각하시는지도 궁금합니다.

✅ **프로그래밍 요구사항에 관하여**

이번 미션의 프로그래밍 요구사항 중에 `함수 들여쓰기 depth를 1로 제한`하는 사항과, `함수의 매개변수는 2개 이하`여야 한다는 사항이 있었습니다.
해당 요구사항은 코드의 가독성을 높이고 함수가 하나의 일을 수행하도록 함수 분리를 요구하는 부분이라고 생각이 들었습니다.

예를 들어, 이번에 구현한 유틸 중 `getUniqueRandomNumbers`라는 메서드는 인자를 3개를 받아서 랜덤 값들을 배열로 반환해주는 로직인데요.
사실 이 로직은 인자 3개가 무조건 필요한 부분이고 순수함수에 가깝기 때문에, `함수의 매개변수는 2개 이하`라는 요구사항을 어떻게 지켜야 할지 고민이 되었습니다.
그렇다고 객체로 3개를 묶어서 보내서 이를 구조분해할당으로 사용해야하는 걸까 싶었습니다.
그렇게 해서라도 매개변수의 개수를 줄여서 프로그래밍 요구사항을 맞추는 게 좋을까요? 아니면 다른 방식이 있을까요?

`함수 들여쓰기 depth를 1로 제한` 요구사항의 경우에도, `forEach` 같은 메서드를 사용하면 들여쓰기가 depth에 1이 추가되는데, 그렇게 되면 금방 함수 들여쓰기 depth 2가 되어버립니다.
그렇다고 이를 또 분리하기에는 오히려 가독성을 더 해칠 수도 있다는 생각이 들었습니다. 너무 비슷한 로직이 분리되면 함수명도 어떻게 지어야할지 더 고민되는 부분이고요.
이렇게 depth를 1로 제한하는 게 과연 맞는 방식일지 궁금합니다. 함수가 하나의 일을 명확히 한다면, 그리고 같이 일하는 개발자분들이 더 이해하기 쉽다면, 그게 더 좋은 코드가 아닐까 생각이 들어요.

✅ **클래스 필드에 관하여**

`함수의 매개변수는 2개 이하`라는 요구사항을 지키기 위해 `발행한 로또 배열(lottos)`를 아예 해당 `WinningStatistics` 클래스 인스턴스를 생성할 때 넣어주도록 수정했습니다.
그렇지 않으면, `calculateWinningResults` 메서드의 인자로 3개의 값이 들어가야 했기 때문입니다.
사실 이 `발행한 로또 배열(lottos)` 변수가 `calculateWinningResults` 내부에서만 사용되는 변수인데, 이렇게 클래스의 필드를 하나 더 추가하여 관리하는 방식이 좀 마음에 들지 않았습니다.
혹시 이런 방식으로 리팩토링 한 것에 대해서 어떻게 생각하시는지 의견을 여쭙고 싶습니다.

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [x] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [x] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [x] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [x] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계

- [ ] 도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [ ] DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가?
- [ ] 웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

sooyeoniya committed:
docs: README.md 기능 요구사항 및 커밋 컨벤션 작성

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(utils): readLineAsync 모듈 추가

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(Input): 구입 금액 입력 구현 및 테스트 작성

- TDD 적용
Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(Input): 당첨 번호 입력 구현 및 테스트 작성

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat: 에러 발생 시 재실행 로직 구현

- throwIfInvalid 유틸 함수 구현
- index.js -> App.js 로 전체 로직 관리 이동

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(Output): 에러 메시지 출력 폼 구현

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(constants): 에러 메시지 상수화 및 적용

- 구입 금액 및 당첨 번호 유효성 검증 에러 메시지 추가
- 테스트 케이스에도 예상 에러 메시지 추가

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
chore: 당첨 번호 유효성 테스트 명세 수정

- 각 테스트 명세에 맞게 문장 수정함
Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(Input & validation): 보너스 번호 입력 및 유효성 검증

- 보너스 번호 입력
- 보너스 번호 유효성 검증
- 테스트 코드 작성 (TDD)

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
chore: 사용하지 않는 import 제거

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
chore: 파일명 복수형으로 수정

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
refactor: 당첨 번호에 있는 빈 값 제거하여 유효성 검증

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
refactor: App.js -> Controller.js 로 변경

- 컨트롤러의 역할을 한다고 판단하여 이름 변경을 통해 역할 명시
Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
refactor: throwIfInvalid 에 추가 인자 받도록 수정

- 보너스 번호 유효성 검증할 때 당첨 번호를 추가 인자로 넣어줘야 하는 경우가 있음
- 재사용성을 위해 추가 인자를 받아 유효성 로직에 넘겨주도록 수정함

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
chore: 안쓰는 import 삭제

sooyeoniya committed:
feat(utils): 중복되지 않는 랜덤 번호 생성 함수 구현

- min, max, count 값을 인자로 받아서 주어진 범위 내에 중복되지 않는 숫자를 count 개수 만큼 추출하여 배열로 반환
Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(issueLottos): 구입 금액 만큼 로또 발행

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(Output): 발행한 로또 출력 로직 구현

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(utils): 두 숫자 배열에서 일치하는 숫자 개수 반환 함수 구현

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
fix(validation): validateBonusNumber에서 보너스 번호 반환

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(WinningStatistics): 발행 로또의 당첨 내역 계산해서 통계내는 클래스 구현

- TDD: WinningStatistics 테스트 코드 작성
- WinningStatistics 클래스 구현 (당첨 번호 계산, 보너스 일치 계산 하여 통계 저장 및 반환)

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
refactor(WinningStatistics): calculateWinningResults 함수 분리

- 함수 하나당 하나의 일을 하도록 수정
- indent depth 줄이기

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(Output): 당첨 통계 내역 출력

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat(WinningStatistics): 당첨 내역에 따라 수익률 계산해서 반환하는 로직 구현

- TDD: 수익률 계산에 대한 테스트 코드 작성
- calculateProfitRatio 메서드 구현
- 수익률 출력 구현

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
feat: 로또 재실행 로직 구현

- TDD: 재실행 입력값 유효성 검증 테스트 코드 작성
- 유효성 검증 로직 구현
- 재실행 로직 구현

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
refactor: let을 const로 변경

- 프로그래밍 요구사항: 변수 const만 사용

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
refactor: 전체 상수화

재사용 되는 매직넘버 상수화하여 constants 파일에 저장
만약 해당 위치에서만 사용되는 매직넘버의 경우 그 내부에서 선언
Input 객체에 있는 입력 프롬프트에 대해서는 응집도를 위해 분리하지 않음

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
refactor: Controller 클래스 함수 길이 제한 적용

- 함수 길이 15 제한으로 적용
- 컨트롤러 함수 길이 제한 과정에서 WinningStatistics 클래스의 calculateWinningResults 메서드의 매개변수 3 -> 2개로 수정
- 매개변수 2개 제한 프로그래밍 요구사항 때문에 줄임.
- 사실 lottos 배열 자체가 calculateWinningResults에서만 사용되지만, 인스턴스 생성할 때 받도록 수정함

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
chore: 유효한 입력 받을때까지 재실행하는 유틸함수 이름 변경
- throwIfInvalid 보다 getValidInput 이 더 적절하다고 판단함
- throwIfInvalid 네이밍은 다른 유틸 함수에서 사용할 예정

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
refactor(validation): 유효성 검증 함수 분리

- 함수 길이 15라인 제한에 따라 함수 분리하여 리팩토링
- throwIfInvalid 유틸 함수를 통해 조건에 따른 에러 반환 로직 재사용

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

sooyeoniya committed:
docs: 페어 회고 작성 및 PR 템플릿 작성

Co-authored-by: AHHYUNJU <AHHYUNJU@users.noreply.github.com>

JUDONGHYEOK commented on src/utils/getUniqueRandomNumbers.js:
```diff
@@ -0,0 +1,9 @@
+const getUniqueRandomNumbers = (min, max, count) => {
+  const randomNumbers = new Set();
+  while (randomNumbers.size < count) {
+    randomNumbers.add(Math.floor(Math.random() * (max - min + 1)) + min);
+  }
+  return Array.from(randomNumbers);
+};
+
+export default getUniqueRandomNumbers;
```
지금도 좋은 로직이지만 최악의 경우에 무한루프에 빠질 가능성이 있어서 아직 시간이 많으니 다른 로직들을 고민해보면 좋을 것 같아요!

sooyeoniya replied:
다른 방식 찾아보고 더 효율적인 방식으로 적용해볼게요!

JUDONGHYEOK replied:
오 저는 index를 기준으로 빼주는 방식으로 생각했는데, 새로운 방식도 좋은 것 같아요! count갯수 이상 즉 필요이상으로 연산이 들어간다는 것을 개선할 수 있을 것 같은데 넘어가도 좋을 것 같습니다!

JUDONGHYEOK commented on src/domains/WinningStatistics.js:
```diff
@@ -0,0 +1,62 @@
+import { MATCH_KEY, MATCH_PRIZE } from "../constants/constants.js";
+class WinningStatistics {
+  #lottos = [];
+  #statistics = new Map([
+    [MATCH_KEY.THREE, { count: 0, amount: MATCH_PRIZE.THREE }],
+    [MATCH_KEY.FOUR, { count: 0, amount: MATCH_PRIZE.FOUR }],
+    [MATCH_KEY.FIVE, { count: 0, amount: MATCH_PRIZE.FIVE }],
+    [
+      MATCH_KEY.FIVE_AND_BONUS,
+      { count: 0, amount: MATCH_PRIZE.FIVE_AND_BONUS },
+    ],
+    [MATCH_KEY.SIX, { count: 0, amount: MATCH_PRIZE.SIX }],
+  ]);
```
여기서 Map을 사용하신 이유가 있을까요? 사용처를 보니 성능상의 이점도 크게 없는 것 같고 가독성 측면에서는 객체가 더 유리한 면들이 있는 것 같아서요!

sooyeoniya replied:
제가 프리코스 당시에 객체로 구현했었는데, 출력하는 과정에서 `5+보너스` 부분과 `6` 부분이 계속 순서가 바뀌어서 출력되더라구요! 그래서 알아보니까 JS 자체 특성때문에 객체를 사용할 경우 속성 순서가 계속 변경되는 문제가 있었어요. 그래서 정수 키 값 먼저 오름차순 정렬되고, 그 뒤에 `5.5`로 되어있는 `MATCH_KEY.FIVE_AND_BONUS` 이 키 값이 마지막에 정렬되더라고요. 애초에 키 값 자체가 숫자로 되어있어서 발생한 문제여서 삽입 순서를 보장해주는 `Map` 객체를 사용하기로 했습니다.

원래는 문자열로 키 값을 정의하고 `if-else`나 `switch` 같은 조건문으로 매칭시켜주는 경우도 생각했는데, 이렇게하면 오히려 `Map` 보다 코드 라인 수가 더 길어져서 가독성을 좀 해친다는 느낌이 들었어요. 하지만 위 `Map` 방식은 상수 이름이 길어져서 오히려 가독성이 좀 떨어지는 느낌도 있긴 하네용.. 그래도 일단 `Map` 객체를 사용해주면 키 값 자체가 숫자 타입이어서 조회나 수정이 빠르다는 장점(`Map`을 쓰면 시간복잡도 측면에서 더 나음)이 있어서 사용하게 되었습니다.

JUDONGHYEOK replied:
제 생각에는 출력과정 자체에서 순서가 중요하다면 출력과정에서 적절하게 키-배열을 활용해서 해결할 수 있을 것  같구요. Map이 객체보다 시간복잡도에서 더 이점을 가져가려면 key-value쌍이 자주 변경되어야된다고 생각하는데 지금은 오히려 key값이 고정되어있어서 그 이점을 가져가지 못한다고 생각했어요.

 가독성의 측면은 아무래도 Map보다 일반객체가 친숙하기도 하고 초기화하는데 있어서 이중배열 구조보다는 계층구조가 더 명확히 포함되기 때문에 일반객체가 더 나은 구조라고 생각했습니다!

 또 오히려 키값이 고정되어 있는 자료구조 형태라 맵구조 보다 단순 해시테이블의 형태가 맞지 않나?라는 생각이 들어서요. 

 사실 큰 차이는 없을 것 같지만 저라면 일반 객체를 사용했을 것 같아서 질문드려보았습니다!!

JUDONGHYEOK replied:
아 자료구조 관련해서는 잘못 코멘트를 달았네요. 자바스크립트 객체 내부에서 해시테이블을 사용해서 성능상의 차이가 크게 나지 않을 거라는 말이였는데 잘못 단 것 같아요. (정확히 해시테이블은 아니고 엔진에 따라 구조 표현이 다르기는 합니다)

JUDONGHYEOK commented on __tests__/WinningStatistics.test.js:
```diff
@@ -0,0 +1,150 @@
+      ]),
+    },
+  ])(
+    "calculateWinningResults 메서드 테스트: $description",
+    ({ lottos, expectedStatistics }) => {
+      // when
+      const winningStatistics = new WinningStatistics(lottos);
+      winningStatistics.calculateWinningResults(winningNumbers, bonusNumber);
+
+      // then
+      expect(winningStatistics.statistics).toEqual(expectedStatistics);
+    },
```
테스트 코드가 전체적으로 each를 통해 중복도 제거해주시고 given-when-then 구조에 description도 알맞게 사용하셔서 가독성이 아주 좋은 것 같아요!!👍

sooyeoniya replied:
칭찬 감사합니다!! 🥹

JUDONGHYEOK commented on __tests__/WinningStatistics.test.js:
```diff
@@ -0,0 +1,150 @@
+    {
+      description: "발행된 로또가 당첨 번호 중 2개와 일치하는 경우",
+      lottos: [[1, 2, 7, 8, 9, 10]],
+      expectedStatistics: new Map([
+        [MATCH_KEY.THREE, { count: 0, amount: MATCH_PRIZE.THREE }],
+        [MATCH_KEY.FOUR, { count: 0, amount: MATCH_PRIZE.FOUR }],
+        [MATCH_KEY.FIVE, { count: 0, amount: MATCH_PRIZE.FIVE }],
+        [
+          MATCH_KEY.FIVE_AND_BONUS,
+          { count: 0, amount: MATCH_PRIZE.FIVE_AND_BONUS },
+        ],
+        [MATCH_KEY.SIX, { count: 0, amount: MATCH_PRIZE.SIX }],
+      ]),
```
Map을 초기화하는 로직이 반복되고 코드에서 많은 라인을 차지 하다는 것 같은데요. 생성 로직을 추상화하는 것은 어떨까요??

sooyeoniya replied:
저도 본 테스트 코드를 작성하는 과정에서 좀 지저분하다는 생각이 들었어요 :) 한 번 적용해보겠습니다 !

JUDONGHYEOK commented on src/views/Output.js:
```diff
@@ -0,0 +1,38 @@
+
+  printStatistics(statistics) {
+    console.log("\n당첨 통계\n--------------------");
+    console.log(
+      `${MATCH_KEY.THREE}개 일치 (${MATCH_PRIZE.THREE.toLocaleString()}원) - ${statistics.get(MATCH_KEY.THREE).count}개`,
+    );
+    console.log(
+      `${MATCH_KEY.FOUR}개 일치 (${MATCH_PRIZE.FOUR.toLocaleString()}원) - ${statistics.get(MATCH_KEY.FOUR).count}개`,
+    );
+    console.log(
+      `${MATCH_KEY.FIVE}개 일치 (${MATCH_PRIZE.FIVE.toLocaleString()}원) - ${statistics.get(MATCH_KEY.FIVE).count}개`,
+    );
+    console.log(
+      `${MATCH_KEY.FIVE}개 일치, 보너스 볼 일치 (${MATCH_PRIZE.FIVE_AND_BONUS.toLocaleString()}원) - ${statistics.get(MATCH_KEY.FIVE_AND_BONUS).count}개`,
+    );
+    console.log(
+      `${MATCH_KEY.SIX}개 일치 (${MATCH_PRIZE.SIX.toLocaleString()}원) - ${statistics.get(MATCH_KEY.SIX).count}개`,
+    );
```
여기도 반복되는 로직이 있는 것 같아요. 객체를 활용해서 추상화하는 것은 어떠세요?

sooyeoniya replied:
한 번 map으로 돌려볼게요 ㅎㅎ! 앞에 상수 부분들을 나중에 추가하다보니 반복 로직이 있는 걸 인지하지 못했나봐요 😃

JUDONGHYEOK commented on src/controllers/Controller.js:
```diff
@@ -0,0 +1,60 @@
+    if (restartConfirm === YES) await this.start();
+  }
+
+  async #runLottoGame() {
```
MVC패턴은 좋은 구조지만 항상 Controller의 과중화나 양방향 데이터흐름 때문에 고민이 되는 것 같아요. 특히나 인터렉션이 많이 발생하는 프론트엔드에서는 Controller가 비대해지는 것을 막는 것도 한계가 있구요. 

때문에 해당 패턴을 채택할 때는  controller를 최대한 가볍게 하려고 노력해야되는 것 같아요. 사실 지금 구조도 충분히 가독성이 좋고 깔끔하지만 유지보수를 위해서 공통적인 로직들을 추후에 또 다른 domain에 옮기거나 validation로직들도 어떤 객체의 책임인지 생각해보아도 좋은 경험이 될 수 있을 것 같습니다!

sooyeoniya replied:
좋은 의견 감사합니다. 제 생각애도 요즘 MVC에 너무 매몰되어서 사고의 틀이 갇혀있다는 생각이 들었어요. 물론 본 미션은 엄청 큰 볼륨은 아니지만, 앞으로 더 좋은 설계를 위해 한 번 고민해보겠습니다!  컨트롤러 최대한 가볍게 할 수 있는 방법으로 고안해서 적용해볼게용 😃

JUDONGHYEOK commented on src/validations/validateWinningNumbers.js:
```diff
@@ -0,0 +1,68 @@
+  );
+};
+
+const chechIsInteger = (winningNumber) => {
```
여기 오타가 있는 것 같습니다!

sooyeoniya replied:
감사합니다 ㅠㅠ 눈이 좀 침침한가봐요 ...ㅎㅎ

JUDONGHYEOK commented on src/validations/validateWinningNumbers.js:
```diff
@@ -0,0 +1,68 @@
+  );
+};
+
+const checkIsNotDuplicate = (winningNumbers, winningNumbersSet) => {
+  throwIfInvalid(
+    winningNumbers.length !== winningNumbersSet.size,
+    WINNING_NUMBERS_ERROR_MESSAGES.DUPLICATE_NUMBER,
+  );
+};
```
winningNumbers와 winningNumbersSet을 동시에 받기 보다는 해당 함수에서 set을 선언하는 것이 해당 함수의 책임에 더 알맞지 않을까요??

sooyeoniya replied:
아 맞네요.. 분리하는 과정에서 왜 같이 넣어줄 생각을 못했는지..!

JUDONGHYEOK commented on __tests__/issueLottos.test.js:
```diff
@@ -0,0 +1,27 @@
+  test("발행된 로또는 오름차순 정렬이 되어 반환된다.", () => {
+    // when
+    const lottos = issueLottos(purchaseAmount);
+    const sortedLottos = lottos.map((lotto) => lotto.sort((a, b) => a - b));
```
js sort()는 원본배열을 직접변경해요. 그래서 lottos가 이차원배열이기만 하면 해당 테스트 코드는 통과되어서 테스트의 의도가 잘 표현되지 않을 것 같아요!

sooyeoniya replied:
헉 알겠습니다! 2차원 배열의 문제점을 제대로 인식하지 못했던 것 같아요.. 불변성 고려해서 2차원 배열 테스트 코드 잘 적용되도록 다시 수정해볼게요 👍🏻

JUDONGHYEOK commented on __tests__/countMatchingNumbers.test.js.js:
```diff
@@ -0,0 +1,31 @@
+      checkingArray: [1, 2],
+      expectedReturnNumber: 2,
+    },
+  ])("$descripion", ({ checkingArray, expectedReturnNumber }) => {
```
여기도 오타가 있네요!

JUDONGHYEOK commented on __tests__/validateBonusNumber.test.js:
```diff
@@ -0,0 +1,46 @@
+    },
+  ])(
+    "보너스 번호가 $description 에러가 발생한다.",
+    ({ input, expectedErrorMessage }) => {
+      // given
+      // when & then
+      expect(() => {
+        const winningNumbers = [11, 12, 13, 14, 15, 16];
+        validateBonusNumber(input, winningNumbers);
+      }).toThrow(expectedErrorMessage);
+    },
```
each 활용을 정말 잘하시는 것 같아요! 정상케이스에 대한 테스트케이스도 추가되면 좋을 것 같습니다!

sooyeoniya replied:
맞는 말씀이십니다. 테스트 커버리지 방금 돌려봤더니 약간 낮게 나온 이유가 정상 케이스에 대한 테스트를 돌리지 않아서 그런 거였네요! 나머지도 동일하게 수정하겠습니다 :)

JUDONGHYEOK commented on __tests__/validateRestartConfirm.test.js:
```diff
@@ -0,0 +1,50 @@
+    }).toThrow(RESTART_ERROR_MESSAGE);
+  });
+
+  test.each([[UPPER_YES, YES, UPPER_NO, NO]])(
```
여기는 2차원배열이라 테스트가 한번만 돌 것 같아요 의도하신 동작은 아니실 것 같아 수정해주시면 좋을 것 같아요!

sooyeoniya replied:
헉 그렇네요! 수정하겠습니다. 감사합니다 !

JUDONGHYEOK reviewed (changes requested):
안녕하세요 니야! 로또미션에서 니야의 리뷰를 맡게된 동키콩입니다!👏  연극은 잘 마치셨나요? 저도 우테코에서 연극을 할 때 극 I로서 미션보다 힘든 경험이였는데요. 또 지나고보니 추억이더라구요. 연극 팀원들이랑도 많이 친해지고 우테코 생활이 좀 더 편해졌던 것 같아요. 니야도 좋은 추억이 되었을거라 믿습니다!👊

 우선 코드를 보고 너무 잘 구조화가 되어있어서 정말 두번째 미션이 맞는지 다시 확인해보았어요. 도메인 로직과 UI로직도 잘 나뉜 것 같고 상수화도 꼼꼼하게 되어있어서 리뷰하기 편한 코드였습니다. 또 테스트 커버리지도 높고 전반적으로 테스트코드도 깔끔하게 추상화하려고 노력하신 모습들이 보여서 인상깊게 보았어요!

 개선할 수 있는 점이라 생각되는 점은 MVC패턴을 채택하여 개발하시다보니 불가피하게 Controller가 비대해지는 문제인데요. 사실 지금 수준에서는 괜찮다고 생각하지만 특히나 MVC패턴을 사용할 때는 최대한 가볍게 만드는 것이 좋다고 생각해서요. 니야가 경험해가면서 해결해 나가실 수 있을 것이라 생각됩니다!

기본질문에 대해서는 코멘트로 남겨두었구요. 나머지 질문들을 보면요.

## TDD관련
> 막상 진행해보니 시간이 생각보다 오래 걸리는 문제 때문에 궁금한 점이 생겼는데요, 마감 기한을 정확히 지켜야하는 실제 실무 상황에서도 TDD를 많이 적용하는 편인지 궁금합니다.
> ...
> 하지만, 꼭 TDD를 레드-그린 리팩토링 방식으로 해야하는지 의문이 들었습니다.
> ...
> 따라서, 추가로 TDD의 효율 및 유용성에 대해서 어떻게 생각하시는지도 궁금합니다.

 TDD에 대해서도 많은 의견들이 나뉘는 것 같아요. 사실 TDD보다 테스트코드 자체가 시간이 많이 걸리는 행위이기 때문에 필요에 따라 적용하는 것 같구요. TDD도 여러 소프트웨어 개발방법론 중 하나라고 생각하여 실무에서는 상황에 맞게 유연하게 적용하는 편인 것 같습니다. 특히나 검증이 확실하게 필요한 부분과 복잡한 비즈니스 로직을 필요로 할때 TDD가 많이 도움이 되는 것 같아요.

 레드-그린리팩토링에 대한 의문이 든 이유는 무엇이였을까요? 도메인 구현 후에 테스트코드를 작성하는 방식은 사실 TDD라고 하기에는 어려워요. 그말은 TDD자체에 의문이 들었다고 여겨지는데요. 사실 지금 요구사항이 미션이다보니 그렇게 복잡하지 않은 도메인이다보니 TDD에 대한 의문이 들 수도 있을 것 같아요. 도메인 구현 후에 테스트코드를 작성하는 방식은 필히 무언가를 놓칠 수 있어요. 특히나 복잡한 도메인이나 이해도가 부족한경우에 많은 것들을 놓치게 됩니다. 코드베이스가 커질수록 그 효용은 크게 다가올 거라 생각이 됩니다. 추가로 [포비의 우아한테크세미나](https://www.youtube.com/watch?v=bIeqAlmNRrA) 를 안보셨다면 강력 추천드립니다!!
 
## 프로그래밍 요구사항
> 그렇다고 객체로 3개를 묶어서 보내서 이를 구조분해할당으로 사용해야하는 걸까 싶었습니다.
> 그렇게 해서라도 매개변수의 개수를 줄여서 프로그래밍 요구사항을 맞추는 게 좋을까요? 아니면 다른 방식이 있을까요?
> 이렇게 depth를 1로 제한하는 게 과연 맞는 방식일지 궁금합니다. 함수가 하나의 일을 명확히 한다면, 그리고 같이 일하는 개발자분들이 더 이해하기 쉽다면, 그게 더 좋은 코드가 아닐까 생각이 들어요.

 매개변수를 2개 이하로 제한하는 것과 depth를 1로 제한하는 이유는 무엇일까요? 매개변수는 특히 실무에서도 3개 이상으로는 잘 설계하지 않는 편이에요. 니야도 말해주었듯이 매개변수가 늘어나면 늘어날수록 사용처에서는 인지적 부담이 증가해요. 특히 동적타입언어인 자바스크립트에서는 최소한의 제한도 사라지게 되죠. 때문에 버그가 발생할 가능성이 늘어나요.

 두가지 제한사항을 둔 이유는 두가지가 무조건적으로 정답이 아니라 의존관계가 복잡하거나 함수의 책임범위를 제한하는 연습을 하는 것이 아닐까 싶어요. 물론 요구사항은 충족해야하지만 니야 나름대로 그것에 대한 정답을 찾았다면 그걸로도 미션의 목표를 달성한 것이 아닐까 생각이 되네요!

## 클래스 필드에 관하여

 저는 오히려 좋은 접근이라고 생각했어요.  lottos라는 필드가 WinningStatistics 객체의 정체성이나 상태를 나타내는 데이터라고 생각했구요. 단순히 함수의 파라미터를 하나 줄이기 위해서의 의도였다면 당연히 좋지 않은 코드이지만 lottos가 WinningStatistics의 상태로 포함되는 것이 객체지향적 설계였다고 한다면 좋은 접근이라고 생각해요. 이번 미션 의도도 그런 상황속에서 객체의 상태나 책임의 객체지향적 설계도 포함된 것이 아닐까도 생각이 되네요.

---
니야! 전체적으로 깔끔하고 니야만의 고민이 많이 들어가 있는 코드라 생각이 되어 좋았습니다. TDD가 처음이라 많이 어색하셨을거라 생각이 되어요. 미션을 떠나 어떻게 하면 TDD가 어떤 문제를 어떻게 해결할 수 있을지 생각해보고 학습해보면 더 재밌게 다가올 수 있을 것 같아요. 현재도 충분히 훌륭한 코드지만 2단계 전 점검할 기회라고 생각이 되어 request chages하도록 하겠습니다! 너무 고생많으셨어요 니야!

sooyeoniya committed:
build: 테스트 커버리지 스크립트 추가

- 생성되는 coverage 폴더 ignore 처리

sooyeoniya committed:
fix(test): 오타 및 테스트케이스 오류 수정

sooyeoniya committed:
test: 정상적인 케이스에 대해 테스트 코드 추가

- 모든 파일에 대하여 테스트 커버리지 100 달성

sooyeoniya committed:
refactor(test): 통계 저장하는 Map 객체 초기화 함수 정의

- Map을 생성하는 로직이 각 테스트 케이스마다 존재하여 코드 라인이 길어지는 이슈가 있었음
- 따라서, Map을 생성하는 초기화 함수를 만들어서 재사용하도록 함

sooyeoniya committed:
refactor: 당첨 통계 내역 객체 순회로 로직 수정

- 원래 하나씩 나열되어있었는데, 객체의 키 값에 따라 반복되는 출력 로직 객체 순회로 변경
- MATCH_PRIZE의 key 값을 MATCH_KEY 의 value 값으로 지정하여 중복되는 키 값 삭제
- 관련 WinningStatistics 메서드 및 테스트 코드 수정

sooyeoniya committed:
chore: 메서드명 오타 수정

- chech -> check

sooyeoniya committed:
refactor: 함수 내부에서 사용되는 로직 위치 안으로 이동

- winningNumbersSet 변수는 checkIsNotDuplicate 내부에서만 사용되는 변수이므로 안으로 위치 옮김

sooyeoniya committed:
refactor: 랜덤 함수 추출 로직 수정

- 기존 코드는 while 문 조건이 충족되지 않을 때, 즉 count 값이 (max - min + 1)보다 큰 경우 무한루프에 빠질 수 있음
- 즉, min과 max 범위 내에서 뽑을 수 있는 숫자의 개수보다 더 많은 숫자를 요구할 경우를 말함
- Set으로 인해 중복제거 되기 때문
- 따라서, min과 max 범위 사이만큼의 모든 숫자를 저장하는 배열 생성 후, 랜덤으로 정렬하여 count 값 만큼 잘라 반환하도록 함

sooyeoniya committed:
fix(test): 오름차순 정렬 테스트 로직 수정

- 기존 테스트 코드에서는 이미 정렬된 lottos를 가지고 테스트 내부에서 다시 정렬하여 테스트하고 있어, 테스트 로직 자체가 잘못되었음
- 따라서, 각 lotto에 하나씩 접근하여 toBeLessThan 매처를 사용해 오름차순인지 하나하나 확인하도록 함

sooyeoniya committed:
refactor(Output): 발행한 로또 출력 로직 매개변수 줄이기

- 발행한 로또의 개수는 purchaseAmount를 1000으로 나누어서 사용하지 않아도, lottos의 길이로 가져올 수 있음
- 이에 따라 purchaseAmount와 MIN_UNIT을 가져오지 않아도 되므로 매개변수 하나로 축소 가능

sooyeoniya committed:
refactor: Controller에서 입력 유효성 검증 메서드 분리

- 컨트롤러가 너무 많은 역할을 부담하고 있음
- 서비스 계층 추가하여 입력 및 유효성 검증 로직 책임 이양
- 서비스 계층 추가한 이유는, InputService 내부 메서드들이 domain과 view가 결합되어있는 로직들이기 때문에
- views에는 두기 어렵다고 판단함. 그러나 Controller에 있기에는 코드라인이 길어짐
- 재사용성이나 유지보수성 측면에서 분리하는 것이 맞다고 판단하여 InputService 를 새로 생성함

sooyeoniya committed:
fix: WinningStatistics 클래스 private 필드 불변성 보장

- 출력 때문에 getter를 사용하고 있는데, 새로운 Map 객체를 반환하도록 하여, 내부 필드가 외부에서 수정되지 않도록 함

sooyeoniya committed:
refactor: Map 객체 생성 함수 utils로 분리

- WinningStatistics 클래스 내부에서도 초기화 하는 로직에 재사용하기 위함

sooyeoniya committed:
refactor: 수익률 천 단위 콤마 추가

- toFixed => toLocaleString으로 변경, 옵션으로 소숫점 표현

sooyeoniya committed:
chore: 파일명과 객체명 통일해주기

- 실수로 객체명 InputReader를 InputView로 수정하지 못함

sooyeoniya commented:
안녕하세요, 동키콩! 니야입니다 :) 주말 잘 보내고 계신가요?! 😁 
저도 처음에는 연극을 대체 왜 하나 싶기도 했지만, 막상 준비하는 과정에서나 연극 발표할 때나 끊임없이 웃으면서 했던 것 같네요 😊 팀원들과 더 깊게 친해질 수 있는 좋은 계기였던 것 같아요! 앞으로도 계속 좋은 추억으로 간직할 수 있을 것 같습니다 ㅎㅎ 

리뷰해주신 내용을 토대로 최대한 반영해봤습니다 ! 아래는 리뷰 반영에 대한 요약표입니다. 

## 📜 리뷰 반영 요약표
| 내용 | 관련 코멘트 | 커밋 |
| ---- | ---------- | ---- | 
| 랜덤 숫자 추출 함수 로직 문제점(무한루프 이슈) | [랜덤 숫자 추출 함수 로직 문제점(무한루프 이슈)](https://github.com/woowacourse/javascript-lotto/pull/345#discussion_r1963450214) | https://github.com/woowacourse/javascript-lotto/pull/345/commits/299237a227dcfe8eed93e275cc3522eb16d7f618 |
| 오타 수정 (1) | [chechIsInteger 오타 ](https://github.com/woowacourse/javascript-lotto/pull/345#discussion_r1963556745) | https://github.com/woowacourse/javascript-lotto/pull/345/commits/0bb9a3cd5266e989646903aafba4ea8c4ddec8d8 |
| 오타 수정 (2) | [descripion 오타](https://github.com/woowacourse/javascript-lotto/pull/345#discussion_r1963573833), [테스트 케이스 오류(2차원 배열)](https://github.com/woowacourse/javascript-lotto/pull/345#discussion_r1963580703) | https://github.com/woowacourse/javascript-lotto/pull/345/commits/051c62c5b441012d3a5ea82546bb15d5ac94b82d |
| 정상 로직 테스트 케이스 추가 | [정상케이스에 대한 테스트케이스도 추가 필요](https://github.com/woowacourse/javascript-lotto/pull/345#discussion_r1963578080) | https://github.com/woowacourse/javascript-lotto/pull/345/commits/087e00a94050f9567dd79b768eacd666417c2aab |
| Map 객체 생성 로직 추상화 | [Map을 초기화하는 로직이 반복](https://github.com/woowacourse/javascript-lotto/pull/345#discussion_r1963518001) | https://github.com/woowacourse/javascript-lotto/pull/345/commits/148decd77f07aeb778eef8055e664d859a3d24bd https://github.com/woowacourse/javascript-lotto/pull/345/commits/7a676cecfdca77eecd5aa8555012cf6040e44d68 |
| 당첨 통계 내역 출력 반복 로직 객체로 순회 | [당첨 통계 내역 출력에 반복되는 로직 존재](https://github.com/woowacourse/javascript-lotto/pull/345#discussion_r1963530488) | https://github.com/woowacourse/javascript-lotto/pull/345/commits/0d397676c3c96daab1717923a98c002c5b997371 |
| 함수 내부에서 사용되는 로직 해당 함수 내부로 이동 | [Set 선언 위치 문제](https://github.com/woowacourse/javascript-lotto/pull/345#discussion_r1963559381) | https://github.com/woowacourse/javascript-lotto/pull/345/commits/2da0bd34120398b9a00e2a917ce8a9aa4e4a89e2 |
| 당첨 번호에 대한 오름차순 정렬 테스트 로직 수정 | [lottos 관련 테스트 의도 불분명 이슈](https://github.com/woowacourse/javascript-lotto/pull/345#discussion_r1963570461) | https://github.com/woowacourse/javascript-lotto/pull/345/commits/c384b8f92a483a53b0f49a38a837e8cad3ec5437|
| Controller 에서 입력 유효성 검증 로직 분리 | [Controller가 비대해지는 문제](https://github.com/woowacourse/javascript-lotto/pull/345#discussion_r1963548328) | https://github.com/woowacourse/javascript-lotto/pull/345/commits/cc72bdf831469947e9568a42c61f9bc78f94a634 https://github.com/woowacourse/javascript-lotto/pull/345/commits/5afb85b8103f421000cbc7ebfa1c0cc596b7424f |

## 📜 기타 커밋 요약표
| 내용 | 커밋 |
| ---- | ---- | 
| 발행한 로또 출력하는 함수의 매개변수 줄이기 | https://github.com/woowacourse/javascript-lotto/pull/345/commits/255406c2c6aab0cb9f6d91bf03c024f05aa980f1 |
| WinningStatistics 클래스 private 필드 불변성 보장 | https://github.com/woowacourse/javascript-lotto/pull/345/commits/617529abecc0518a9945a6c052f9d743584ba61f |
| 수익률 천 단위 콤마 추가하여 출력 | https://github.com/woowacourse/javascript-lotto/pull/345/commits/5444305a2ef7d8c289a6ca0012bae72350151fbf |

## ❓궁금한 점
### `InputService` 의 파일 위치에 대해 궁금한 점이 있습니다. 
본 객체 내부에 있는 메서드들은 모두 도메인과 뷰가 조합되어 있는 메서드들입니다. 이 메서드들을 컨트롤러에서 분리하려다보니 뷰에 위치하기에는 도메인 로직이 섞여있어 적절하지 못하다고 판단했습니다. 따라서 `Service` 라는 새로운 계층을 생성해 추가해주었습니다. 제가 생각한 본 계층의 역할은 컨트롤러와 뷰, 그리고 도메인의 오작교 역할을 해주는 그런 중간다리 계층이라고 생각했습니다. 즉 사용자로부터 입력 값을 받아와 그 값의 유효성 검증을 위해 도메인을 부르고, 검증된 값을 컨트롤러에 전달해주는 것입니다. 이렇게 분리하면 재사용성이나 유지보수성 측면에서 훨씬 유용한 방식이긴 하다고 생각했습니다.

그러나 코드를 보다보니 폴더 계층 구조에서 `Service` 가 과연 본 객체가 존재하기에 적절한 위치인지 궁금해졌습니다. 단순히 컨트롤러의 볼륨을 줄이려다보니 생겨난 계층인데, 이렇게 계층을 새로 생성하지 않고, 아예 새로운 두 번째 컨트롤러(입력 및 검증만 담당)를 생성하는 게 더 깔끔한 방식이었을지 고민이 됩니다. 왜냐하면, 어쨌든 본 객체는 도메인과 뷰가 조합된 메서드들을 가지고 있기 때문에, 사실상 도메인과 뷰를 상호작용해주는 컨트롤러의 역할이 맞기 때문이죠.! 굳이 `Service` 계층을 만들어서 더 계층 구조가 복잡해진건 아닐까 하는 생각이 듭니다.

어디에 위치시키는 게 더 괜찮은 구조로 보이는지 동키콩의 의견도 궁금합니다!

JUDONGHYEOK reviewed (approved):
안녕하세요 니야! 와우 리뷰반영해주신 것을 표로 이렇게 깔끔하게 정리해주셔서 빠르게 파악하고 확인할 수 있었어요. 많은 리뷰들을 대부분 반영해주신 것 같고 모든 요구사항을 만족하고 있다고 생각해서 어프로브 하도록 하겠습니다!

사실 제가 컨트롤러 관련 리뷰를 한 의도는 도메인 로직들을 묶을 수 있는 중간 도메인을 추가로 도입해보면 어떻겠냐는 생각이였어요. 하지만 현재 코드와 같이 니야 나름대로 고민하고 새로운 계층을 추가해주신 것도 좋은 접근법이라고 생각해요.

 제가 잘못 이해한 것일 수도 있지만 InputService가 또 다른 Controller에서 의존하고 있기 때문에 Controller와는 다른 계층의 역할을 하는 것이 아닐까 생각하게 된 것 같은데요. 그런데 현재 코드만 보면 inputView와 validate(도메인성)를 연결해주는 역할을 하는 객체로 보여요. 따라서 제 생각에는 Service 계층보다는 입력과 검증을 담당하는 별도의 Controller라고 보는 것이 더 적절할 것 같습니다!!

1단계 리뷰도 빠르게 반영해주셔서 감사해요 니야! 니야 코드보면서 저도 많이 배웠던 것 같아요! 2단계 미션도 화이팅입니다~~👏🦍
